### PR TITLE
Call init's parameter interaction before we create the UI options model

### LINKED ISSUE #6773: -connect doesn't turn off -listen in bitcoin-qt
I built commit a75c67364d3484147c0643baf535845ce72a50cc (on the master branch), dated Mon Oct 5 13:41:27 2015 +0200.

When I run `bitcoind`, I can specify `-connect=127.0.0.1` and it turns off the `-listen` option automatically.

If I try the same with `bitcoin-qt` it doesn't.

`qt/optionsmodel.cpp` is running the following code to set `-listen` to its default value before any of the logic about `-connect` implying `-listen=0` runs:

    if (!SoftSetBoolArg("-listen", settings.value("fListen").toBool()))
        addOverriddenOption("-listen");

Later, `init.cpp` runs:

        if (SoftSetBoolArg("-listen", false))
            LogPrintf("%s: parameter interaction: -connect set -> setting -listen=0\n", __func__);

but it's too late. `SoftSetBoolArg` will only change the value of the setting once, and `optionsmodel.cpp` has already set it.

### PR DESCRIPTION
**Pull request by jonasschnelli**
Currently `optionsModel`(QT) parameters interaction overwrites/dominates initial and depending parameter settings. This PR would factor out init's parameter interaction and call it before `AppInit2()`.

Fixes #6773

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Force pushed a copy'n'paste fix: `s/bool InitParameterInteraction()/void InitParameterInteraction()`.
Credits to @MarcoFalke on IRC.

**Comment by MarcoFalke** [comment]
Concept ACK.

**Comment by laanwj** [comment]
Concept ACK

**Comment by paveljanik** [comment]
ACK

File: src/qt/bitcoin.cpp [hunk 1]
```diff
@@ -381,7 +398,7 @@ void BitcoinApplication::startThread()
     if(coreThread)
         return;
     coreThread = new QThread(this);
-    BitcoinCore *executor = new BitcoinCore();
+    executor = new BitcoinCore();
     executor->moveToThread(coreThread);
```
Comment (laanwj) [inline src/qt/bitcoin.cpp]: It's not allowed to use this object after moving it to another thread.
I'd suggest calling `InitParameterInteraction()` directly from `BitcoinApplication::parameterSetup()` and not bothering with the stuff in BitcoinCore.

**Comment by jonasschnelli** [comment]
Rebased and fixed the pointer access after `moveToThread` by following @laanwj advice.

**Comment by gmaxwell** [comment]
Sorry, blocksonly mode caused a need for a small change here, as it needs to get moved. @jonasschnelli  mind rebasing?

What about that prune/txindex "interaction" (fails to run), should it move too?

**Comment by jonasschnelli** [comment]
> Sorry, blocksonly mode caused a need for a small change here, as it needs to get moved. @jonasschnelli mind rebasing?

Just rebased and moved the `-blocksonly` interaction to the new function.

> What about that prune/txindex "interaction" (fails to run), should it move too?

My idea was to only move "rules" to the new function. Things that evaluate parameters through `GetBoolArg` or `mapArgs`, etc. and depending on its state, optional change another startup variable through `SoftSetBoolArg`, etc.

Pruning and TxIndex are checks that lead to a `InitError()` (startup termination) and it would require to move the `InitError()` context to the `ParameterInteraction()` function.

**Comment by MarcoFalke** [comment]
utACK 51627cc

**Comment by jonasschnelli** [comment]
Would be nice if someone finds time to test this. Binaries: https://bitcoin.jonasschnelli.ch/pulls/6780/
This fixes a ugly bug (#6773) which would be nice to have fixed in 0.12.

**Comment by fanquake** [comment]
Will do.

**Comment by MarcoFalke** [comment]
@jonasschnelli 

* master: 

```
$ src/bitcoind -regtest -connect=127.0.0.1 -server -printtoconsole |grep parameter 
2015-11-21 17:52:56 AppInit2: parameter interaction: -connect set -> setting -dnsseed=0
2015-11-21 17:52:56 AppInit2: parameter interaction: -connect set -> setting -listen=0
2015-11-21 17:52:56 AppInit2: parameter interaction: -listen=0 -> setting -upnp=0
2015-11-21 17:52:56 AppInit2: parameter interaction: -listen=0 -> setting -discover=0
2015-11-21 17:52:56 AppInit2: parameter interaction: -listen=0 -> setting -listenonion=0

$ src/qt/bitcoin-qt -regtest -connect=127.0.0.1 -server -printtoconsole |grep parameter 
2015-11-21 17:53:34 AppInit2: parameter interaction: -connect set -> setting -dnsseed=0
```

* 51627cc

```
$ ~/bitcoin* -regtest -connect=127.0.0.1 -debug=1 -printtoconsole |grep parameter 
```

**Comment by dcousens** [comment]
concept ACK / utACK

**Comment by jonasschnelli** [comment]
I'd like to take this into 0.12 because it fixes a slightly security-critical issue (#6773). Some tested ACKs would be nice (another 5 days left until feature freeze).

**Comment by MarcoFalke** [comment]
@jonasschnelli I am happy to test but the debug.log issue seems not yet resolved? Am I missing something?

**Comment by laanwj** [comment]
utACK. Agree it needs to be in 0.12

Haven't tested any specific cases - I've been holding off on this because the parameter processing is kind of fragile and I haven't properly reasoned it through yet that this doesn't have effect on the precedence of options specified in e.g. configuration files, command line, QSettings.

**Comment by jonasschnelli** [comment]
Thanks @MarcoFalke for pointing out the logging issue. Added another commit that moves the logging initialization.

For reviewers: The critical part are the first 100 lines in `AppInit2()` because the logging init and the parameter interactions are now before `AppInit2()`.

**Comment by MarcoFalke** [comment]
Sanity tests pass now:

```
$ src/bitcoind -regtest -connect=127.0.0.1 -server -printtoconsole |egrep "(parameter|Bitcoin)"
2015-11-27 07:39:08 Bitcoin version v0.11.99.0-296c42e (2015-11-26 14:03:27 +0100)
2015-11-27 07:39:08 InitParameterInteraction: parameter interaction: -connect set -> setting -dnsseed=0
2015-11-27 07:39:08 InitParameterInteraction: parameter interaction: -connect set -> setting -listen=0
2015-11-27 07:39:08 InitParameterInteraction: parameter interaction: -listen=0 -> setting -upnp=0
2015-11-27 07:39:08 InitParameterInteraction: parameter interaction: -listen=0 -> setting -discover=0
2015-11-27 07:39:08 InitParameterInteraction: parameter interaction: -listen=0 -> setting -listenonion=0

$ src/qt/bitcoin-qt -regtest -connect=127.0.0.1 -server -printtoconsole |egrep "(parameter|Bitcoin)"
2015-11-27 07:53:02 Bitcoin version v0.11.99.0-296c42e (2015-11-26 14:03:27 +0100)
2015-11-27 07:53:02 InitParameterInteraction: parameter interaction: -connect set -> setting -dnsseed=0
2015-11-27 07:53:02 InitParameterInteraction: parameter interaction: -connect set -> setting -listen=0
2015-11-27 07:53:02 InitParameterInteraction: parameter interaction: -listen=0 -> setting -upnp=0
2015-11-27 07:53:02 InitParameterInteraction: parameter interaction: -listen=0 -> setting -discover=0
2015-11-27 07:53:02 InitParameterInteraction: parameter interaction: -listen=0 -> setting -listenonion=0

**Comment by MarcoFalke** [comment]
utACK 296c42e

File: src/init.cpp [hunk 2]
```diff
@@ -821,16 +839,6 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
...
-    // disable walletbroadcast and whitelistalwaysrelay in blocksonly mode
-    if (GetBoolArg("-blocksonly", DEFAULT_BLOCKSONLY)) {
-        if (SoftSetBoolArg("-whitelistalwaysrelay", false))
-            LogPrintf("%s: parameter interaction: -blocksonly=1 -> setting -whitelistalwaysrelay=0\n", __func__);
-#ifdef ENABLE_WALLET
-        if (SoftSetBoolArg("-walletbroadcast", false))
-            LogPrintf("%s: parameter interaction: -blocksonly=1 -> setting -walletbroadcast=0\n", __func__);
-#endif
-    }
-
```
Comment (MarcoFalke) [inline src/init.cpp]: @jonasschnelli Trivial white space rebase conflict. (Looks like you can ignore changes made in master)

**Comment by jonasschnelli** [comment]
Thanks @MarcoFalke: rebased!

**Pull request merged**